src/internal/pachd/builder.go

```diff
@@ -0,0 +1,397 @@
+	}
+	b.env = serviceenv.InitWithKube(serviceenv.NewConfiguration(config))
+	if b.env.Config().LogFormat == "text" {
+		log.SetFormatter(logutil.FormatterFunc(logutil.Pretty))
+	}
+	profileutil.StartCloudProfiler(service, b.env.Config())
+	debug.SetGCPercent(b.env.Config().GCPercent)
+	if b.env.Config().EtcdPrefix == "" {
+		b.env.Config().EtcdPrefix = collection.DefaultPrefix
+	}
+
+	b.authInterceptor = authmw.NewInterceptor(b.env.AuthServer)
+	b.loggingInterceptor = loggingmw.NewLoggingInterceptor(b.env.Logger())
+	b.txnEnv = transactionenv.New()
+
+	return
+}
+
+func (b *builder) setupDB(ctx context.Context) error {
+	// TODO: currently all pachds attempt to apply migrations, we should coordinate this
+	if err := dbutil.WaitUntilReady(context.Background(), log.StandardLogger(), b.env.GetDBClient()); err != nil {
+		return err
+	}
+	if err := migrations.ApplyMigrations(context.Background(), b.env.GetDBClient(), migrations.MakeEnv(nil, b.env.GetEtcdClient()), clusterstate.DesiredClusterState); err != nil {
+		return err
+	}
+	if err := migrations.BlockUntil(context.Background(), b.env.GetDBClient(), clusterstate.DesiredClusterState); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *builder) initDexDB(ctx context.Context) error {
+	b.env.InitDexDB()
+	return nil
+}
+
+func (b *builder) maybeInitReporter(ctx context.Context) error {
+	if b.env.Config().Metrics {
+		b.reporter = metrics.NewReporter(b.env)
+	}
+	return nil
+}
+
+func (b *builder) initAuthInterceptor() {
+	b.authInterceptor = authmw.NewInterceptor(b.env.AuthServer)
+}
+
+func (b *builder) initLoggingInterceptor() {
+	b.loggingInterceptor = loggingmw.NewLoggingInterceptor(b.env.Logger())
+}
+
+func (b *builder) initInternalServer(ctx context.Context) error {
+	var err error
+	if b.authInterceptor == nil {
+		b.initAuthInterceptor()
+	}
+	if b.loggingInterceptor == nil {
+		b.initLoggingInterceptor()
+	}
+	b.daemon.internal, err = grpcutil.NewServer(
+		ctx,
+		false,
+		grpc.ChainUnaryInterceptor(
+			errorsmw.UnaryServerInterceptor,
+			tracing.UnaryServerInterceptor(),
+			b.authInterceptor.InterceptUnary,
+			b.loggingInterceptor.UnaryServerInterceptor,
+		),
+		grpc.ChainStreamInterceptor(
+			errorsmw.StreamServerInterceptor,
+			tracing.StreamServerInterceptor(),
+			b.authInterceptor.InterceptStream,
+			b.loggingInterceptor.StreamServerInterceptor,
+		),
+	)
+	return err
+}
+
+func (b *builder) initExternalServer(ctx context.Context) error {
+	var err error
+	if b.authInterceptor == nil {
+		b.initAuthInterceptor()
+	}
+	if b.loggingInterceptor == nil {
+		b.initLoggingInterceptor()
+	}
+	b.daemon.external, err = grpcutil.NewServer(
+		ctx,
+		true,
+		// Add an UnknownServiceHandler to catch the case where the user has a client with the wrong major version.
+		// Weirdly, GRPC seems to run the interceptor stack before the UnknownServiceHandler, so this is never called
+		// (because the version_middleware interceptor throws an error, or the auth interceptor does).
```

Comment:
I think I might not fully understand this comment. If the unknownServiceHandler is never called, why are we adding it since it won't be able to catch the major upgrade error, right (because its never called)?

Reply:
I don’t know; it dates back to [this comment](https://github.com/pachyderm/pachyderm/commit/5ab8a0f7e632a31d41d080cc89913c1e262be2d6#diff-1aa4ab905e56432d62cd27077bfb5fbd35430854b2cbaab1de9cf2603e294c10R144-R160). Unfortunately Alysha is no longer here. Perhaps @acohen4 has some context?

Reply:
I'm not sure - I agree it seems like it can be removed, but if we do so it should be in a follow up PR, so that this one can be interpreted as closer to a true refactor